mlflow/pyspark/ml/__init__.py

```diff
@@ -107,24 +120,129 @@ def _get_estimator_info_tags(estimator):
+    return isinstance(instance, (CrossValidatorModel, TrainValidationSplitModel))
+
+
+def _should_log_hierarchy(estimator):
+    from pyspark.ml import Pipeline
+    from pyspark.ml.classification import OneVsRest
+    return isinstance(estimator, (Pipeline, OneVsRest)) or \
+        _is_parameter_search_estimator(estimator)
+
+
+AutologgingEstimatorMetadata = namedtuple('AutologgingEstimatorMetadata', [
+    'hierarchy', 'uid2name_map', 'param_search_estimators'
+])
+
+
+def _gen_instance_metadata_recursively(
+        instance,
+        uid2name_map,
+        cls2count_map,
+        cls2first_uid_map,
+        param_search_estimators,
+):
     from pyspark.ml import Pipeline
+    from pyspark.ml.classification import OneVsRest
+
+    def gen_instance_name(instance):
+        stage_cls = instance.__class__.__name__
+        stage_uid = instance.uid
 
-    stage_hierarchy = []
-    pipeline_stages = pipeline.getStages()
-    for stage in pipeline_stages:
-        if isinstance(stage, Pipeline):
-            hierarchy_elem = _get_pipeline_stage_hierarchy(stage)
+        if stage_cls not in cls2count_map:
+            cls2count_map[stage_cls] = 1
+            uid2name_map[stage_uid] = stage_cls
+            cls2first_uid_map[stage_cls] = stage_uid
         else:
-            hierarchy_elem = stage.uid
-        stage_hierarchy.append(hierarchy_elem)
-    return {pipeline.uid: stage_hierarchy}
+            cls2count_map[stage_cls] += 1
+            uid2name_map[stage_uid] = f'{stage_cls}_{cls2count_map[stage_cls]}'
+
+        return uid2name_map[stage_uid]
+
+    instance_name = gen_instance_name(instance)
+    if isinstance(instance, Pipeline):
+        sub_stages = []
+        for sub_stage in instance.getStages():
+            sub_stages.append(_gen_instance_metadata_recursively(
+                sub_stage, uid2name_map, cls2count_map, cls2first_uid_map, param_search_estimators
+            ))
+        return {'name': instance_name, 'stages': sub_stages}
+    elif isinstance(instance, OneVsRest):
+        classifier = _gen_instance_metadata_recursively(
+            instance.getClassifier(),
+            uid2name_map, cls2count_map, cls2first_uid_map, param_search_estimators
+        )
+        return {'name': instance_name, 'classifier': classifier}
+    elif _is_parameter_search_estimator(instance):
+        param_search_estimators.append(instance)
+        evaluator = instance.getEvaluator()
+        tuned_estimator = instance.getEstimator()
+        return {
+            'name': instance_name,
+            'evaluator': _gen_instance_metadata_recursively(
+                evaluator, uid2name_map, cls2count_map, cls2first_uid_map, param_search_estimators
+            ),
+            'tuned_estimator': _gen_instance_metadata_recursively(
+                tuned_estimator, uid2name_map, cls2count_map, cls2first_uid_map,
+                param_search_estimators
+            )
+        }
+    else:
+        return instance_name
```

Comment:
```suggestion
        return {"name': instance_name}
```

Should we return a dict here to make all items in the hierarchy dict have the same format (`{"name": "xyz", ...}`)

```jsonc
{
  # current
  "hierarchy": {
    "name": "CrossValidator",
    "evaluator": "BinaryClassificationEvaluator",
    "tuned_estimator": {
      "name": "Pipeline",
      "stages": [
        "Tokenizer",
        "HashingTF",
        "LogisticRegression"
      ]
    }
  },
  # proposed?
  "hierarchy": {
    "name": "CrossValidator",
    "evaluator": "BinaryClassificationEvaluator",
    "tuned_estimator": {
      "name": "Pipeline",
      "stages": [
        {"name": "Tokenizer"},
        {"name": "HashingTF"},
        {"name": "LogisticRegression"}
      ]
    }
  }
}

```

Reply:
yes. make sense.